apps/api-extractor/src/collector/MessageRouter.ts

```diff
@@ -0,0 +1,374 @@
+} from '../api/IExtractorConfig';
+import { AedocDefinitions } from '../aedoc/AedocDefinitions';
+import { ILogger } from '../api/ILogger';
+
+interface IReportingRule {
+  logLevel: ExtractorMessageLogLevel;
+  addToApiReviewFile: boolean;
+}
+
+export class MessageRouter {
+  // All messages
+  private readonly _messages: ExtractorMessage[];
+
+  // For each AstDeclaration, the messages associated with it.  This is used when addToApiReviewFile=true
+  private readonly _associatedMessagesForAstDeclaration: Map<AstDeclaration, ExtractorMessage[]>;
+
+  // Messages that got written to the API review file
+  private readonly _messagesAddedToApiReviewFile: Set<ExtractorMessage>;
+
+  // Normalized representation of the routing rules from api-extractor.json
+  private _reportingRuleByMessageId: Map<string, IReportingRule> = new Map<string, IReportingRule>();
+  private _compilerDefaultRule: IReportingRule = { logLevel: ExtractorMessageLogLevel.None,
+    addToApiReviewFile: false };
+  private _extractorDefaultRule: IReportingRule = { logLevel: ExtractorMessageLogLevel.None,
+    addToApiReviewFile: false };
+  private _tsdocDefaultRule: IReportingRule = { logLevel: ExtractorMessageLogLevel.None,
+    addToApiReviewFile: false };
+
+  public constructor(messagesConfig: IExtractorMessagesConfig) {
+    this._messages = [];
+    this._associatedMessagesForAstDeclaration = new Map<AstDeclaration, ExtractorMessage[]>();
+    this._messagesAddedToApiReviewFile = new Set<ExtractorMessage>();
+
+    this._applyMessagesConfig(messagesConfig);
+  }
+
+  /**
+   * Read the api-extractor.json configuration and build up the tables of routing rules.
+   */
+  private _applyMessagesConfig(messagesConfig: IExtractorMessagesConfig): void {
+    if (messagesConfig.compilerMessageReporting) {
+      for (const messageId of Object.getOwnPropertyNames(messagesConfig.compilerMessageReporting)) {
+        const reportingRule: IReportingRule = MessageRouter._getNormalizedRule(
+          messagesConfig.compilerMessageReporting[messageId]);
+
+        if (messageId === 'default') {
+          this._compilerDefaultRule = reportingRule;
+        } else if (!/^TS[0-9]+$/.test(messageId)) {
+          throw new Error(`Error in API Extractor config: The messages.compilerMessageReporting table contains`
+            + ` an invalid entry "${messageId}". The identifier format is "TS" followed by an integer.`);
+        } else {
+          this._reportingRuleByMessageId.set(messageId, reportingRule);
+        }
+      }
+    }
+
+    if (messagesConfig.extractorMessageReporting) {
+      for (const messageId of Object.getOwnPropertyNames(messagesConfig.extractorMessageReporting)) {
+        const reportingRule: IReportingRule = MessageRouter._getNormalizedRule(
+          messagesConfig.extractorMessageReporting[messageId]);
+
+        if (messageId === 'default') {
+          this._extractorDefaultRule = reportingRule;
+        } else if (!/^ae-/.test(messageId)) {
```

Comment:
Is this testing for AE's own messages? This seems wrong. Can we attach some richer metadata to the messages?

Reply:
The **api-extractor.json** config file has 3 separate rule tables:  `compilerMessageReporting`,  `extractorMessageReporting`, and `tsdocMessageReporting`.  The key for each section follows a recognizable naming pattern (`TS1234` for the compiler, `ae-something` for extractor, and `tsdoc-something` for the TSDoc issues).

In order to provide more informative error messages, we perform validation in two steps:
1. Does the ID follow the naming pattern?  If not, maybe people are putting the rules in the wrong table. In this case the error message is telling you "hey, the code doesn't look right." OTHERWISE,
2. Is it a known ID?  For example, `ae-blah` does look like an API Extractor error, but it's not a code that we recognize.  So maybe they misspelled it.  Or maybe the ID is from a different version of the tool.